### PR TITLE
Add the flat-s hist hook as option for histograms

### PR DESCRIPTION
**Pull request by Bogdan-Wiederspan**
Within our histogram pipeline no rebinning is currently possible. This kind of feature is useful for inference models in combine.

    implements a hist-hook called "flat-s" that takes a finely binned histogram and rebins the histogram for given constraints.
    The hist-hook is split in 2 parts:
        - finding edges that full-fill the constraints
        - apply edges on given "Hist" histograms

### CODE REVIEW

File: hbt/config/configs_hbt.py [hunk 1]
```diff
@@ -147,6 +147,10 @@ def if_era(
         if process_name.startswith(("graviton_hh_", "radion_hh_")):
             proc.add_tag("signal")
             proc.add_tag("resonant_signal")
+        if process_name.startswith("tt"):
+            proc.add_tag("is_ttbar")
```
Comment (riga) [inline hbt/config/configs_hbt.py:151]: Can you rename these tags to "ttbar" and "dy" (to be consistent with other tags already defined above).

File: hbt/config/hist_hooks.py [hunk 2]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
 
         return hists
 
+    def flat_s(task, hists: dict[hist.Histogram]) -> dict[hist.Histogram]:
```
Comment (riga) [inline hbt/config/hist_hooks.py:190]: `dict[hist.Histogram]` only refers to the dict values so it's invalid.

File: hbt/config/hist_hooks.py [hunk 3]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+        """Rebinnig of the histograms in *hists* to archieve a flat-signal distribution.
+        Args:
+
+            task (TODO): task instance that contains the process informations
+            hists (Dict[hist.Histogram]): A dictionary of histograms using Process instances as keys
+
+        Returns:
+            Dict[hist.Histogram]: A dictionary of histograms using Process instances as keys
+        """
```
Comment (riga) [inline hbt/config/hist_hooks.py:199]: Can you adapt the docs style to what is used in cf?

File: hbt/config/hist_hooks.py [hunk 4]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+            # calculate cumulative of reversed signal yield and yield per bin
+            cumulu_y_signal = np.cumsum(y_signal, axis=0)
+            full_cum = cumulu_y_signal[-1]
+            y_per_bin = full_cum / n_bins
+            num_events = len(cumulu_y_signal)
+
+            # prepare background
+
+            for process, histogram in background_histograms.items():
+                if process.name == "tt":
```
Comment (riga) [inline hbt/config/hist_hooks.py:261]: Use the tags you defined yourself ;)

File: hbt/config/hist_hooks.py [hunk 5]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+                    next_idx = offset + np.where(cumulu_y_signal[offset:] > threshold)[0][0]
+                else:
+                    # special case: remaining signal yield smaller than the expected per-bin yield,
+                    # so find the last event
+                    next_idx = offset + np.where(cumulu_y_signal[offset:])[0][-1] + 1
+
+                # advance the index until backgrounds constraints are met
+
+                # combine tt and dy histograms
+
```
Comment (riga) [inline hbt/config/hist_hooks.py:295]: ?

File: hbt/config/hist_hooks.py [hunk 6]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+                bin_edges.append(max(min(edge_value, max_edge), low_edge))
+
+                y_already_binned += get_integral(cumulu_y_signal, next_idx, offset)
+                offset = next_idx
+                indices_gathering.append(next_idx)
+
+            # make sure the lower dnn_output (max events) is included
+            if bin_edges[-1] != low_edge:
+                if len(bin_edges) > n_bins:
+                    raise RuntimeError(f"number of bins reached and initial bin edge is not minimal bin edge (edges: {bin_edges})") # noqa
```
Comment (riga) [inline hbt/config/hist_hooks.py:352]: Use two lines and indent properly instead of `# noqa` instructions.

File: hbt/config/hist_hooks.py [hunk 7]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+                    f"    -> reason: {stop_reason or 'NO REASON!?'}",
+                )
+                n_bins = n_bins_actual
+
+            # flip indices to the right order
+            indices_gathering = (np.flip(indices_gathering) - num_events) * -1
+            return np.flip(np.array(bin_edges), axis=-1), indices_gathering
+
+        def apply_edges(h: hist.Hist, edges: np.ndarray, indices: np.ndarray, variable: tuple[str]) -> hist.Hist:
+            """
```
Comment (riga) [inline hbt/config/hist_hooks.py:372]: Adapt doc style, see above.

File: hbt/config/hist_hooks.py [hunk 8]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+                    f"  started from {num_events} bins, targeted {n_bins} but ended at {n_bins_actual} bins\n"
+                    f"    -> reason: {stop_reason or 'NO REASON!?'}",
+                )
+                n_bins = n_bins_actual
+
+            # flip indices to the right order
+            indices_gathering = (np.flip(indices_gathering) - num_events) * -1
+            return np.flip(np.array(bin_edges), axis=-1), indices_gathering
+
+        def apply_edges(h: hist.Hist, edges: np.ndarray, indices: np.ndarray, variable: tuple[str]) -> hist.Hist:
```
Comment (riga) [inline hbt/config/hist_hooks.py:371]: `variable` (singular), but `tuple[str]`?

File: hbt/config/hist_hooks.py [hunk 9]
```diff
@@ -187,6 +187,278 @@ def qcd_estimation(task, hists):
...
+            """
+            Rebin the content axes determined by *variables* of a given hist histogram *h* to
+            given *edges* and their *indices*.
+            The rebinned histogram is returned.
+
+            Args:
+            h (hist.Hist): hist Histogram that is to be rebinned
+            edges (np.ndarray): a array of ascending bin edges
+            indices (np.ndarray): a array of indices that define the new bin edges
+            variables (str): variable name that is rebinned
```
Comment (riga) [inline hbt/config/hist_hooks.py:381]: `variables` (plural) but `str`?

**Review by riga — CHANGES_REQUESTED** [review]
Thanks for porting this!
Just a couple inline comments that should be resolved imho.

File: hbt/config/configs_hbt.py (hunk 1 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/configs_hbt.py:151]: done

File: hbt/config/hist_hooks.py (hunk 4 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:261]: done, now using `ttbar`

File: hbt/config/hist_hooks.py (hunk 6 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:352]: is splitted into 2 lines

File: hbt/config/hist_hooks.py (hunk 8 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:371]: is only str, nut tuple of strings. Is changed accordingly

File: hbt/config/hist_hooks.py (hunk 9 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:381]: is singular - see change above.

File: hbt/config/hist_hooks.py (hunk 3 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:199]: done, please look if the new doc style is right. Should I also add raises into docstring?

File: hbt/config/hist_hooks.py (hunk 7 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py:372]: see above

File: hbt/config/hist_hooks.py (hunk 2 above)
Comment (Bogdan-Wiederspan) [reply hbt/config/hist_hooks.py]: done

File: hbt/config/hist_hooks.py [hunk 10]
```diff
@@ -187,6 +187,291 @@ def qcd_estimation(task, hists):
...
+                y_already_binned += get_integral(cumulu_y_signal, next_idx, offset)
+                offset = next_idx
+                indices_gathering.append(next_idx)
+
+            # make sure the lower dnn_output (max events) is included
+            if bin_edges[-1] != low_edge:
+                if len(bin_edges) > n_bins:
+                    raise RuntimeError(
+                        "number of bins reached and initial bin edge"
+                        f"is not minimal bin edge (edges: {bin_edges})")
```
Comment (riga) [inline hbt/config/hist_hooks.py:366]:
```suggestion
                    raise RuntimeError(
                        "number of bins reached and initial bin edge"
                        f" is not minimal bin edge (edges: {bin_edges})",
                    )
```

File: hbt/config/hist_hooks.py [hunk 11]
```diff
@@ -187,6 +187,291 @@ def qcd_estimation(task, hists):
...
+            indices_gathering = (np.flip(indices_gathering) - num_events) * -1
+            return np.flip(np.array(bin_edges), axis=-1), indices_gathering
+
+        def apply_edges(h: hist.Hist, edges: np.ndarray, indices: np.ndarray, variable: str) -> hist.Hist:
+            """
+            Rebin the content axes determined by *variable* of a given hist histogram *h* to
+            given *edges* and their *indices*.
+            The rebinned histogram is returned.
+
+
```
Comment (riga) [inline hbt/config/hist_hooks.py:391]: Stray line.

**Review by riga — CHANGES_REQUESTED** [review]
Thanks for the follow-up. I just have three residual comments that should be pretty easy to resolve.

**Review by riga — APPROVED** [review]

**Pull request merged**